In [5]:
import pandas as pd
import numpy as np
import os
import pickle
import time
import optuna
import warnings

from category_encoders import TargetEncoder
from xgboost import XGBRegressor
from scipy.stats import randint, uniform
from lightgbm import LGBMRegressor


from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import Lasso, LinearRegression
from sklearn.model_selection import train_test_split, cross_val_score, RandomizedSearchCV, _search_successive_halving, GridSearchCV
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, ExtraTreesRegressor, HistGradientBoostingRegressor 
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error

warnings.filterwarnings("ignore")


In [6]:
cleanData = pd.read_csv('./data csv/kijkcijfersWeer.csv')
cleanData.sample(5)

,Unnamed: 0,FullDate,date,hour,Kanaal,Programma,Lengte_sec,Kijkers,Temperatuur,Gevoelstemp,Regen,Sneeuw,Weercode,Bewolking,Windsnelheid,Zonnenschijn,isFeestdag,Weekdag,isWeekend,Seizoen
26787,26787,2020-06-08 22:40:41,2020-06-08,22,EEN,HET JOURNAAL LAAT,1080,416865,12.4,9.9,0.0,0.0,3,98,14.5,0.0,0,0,0,lente
27677,27677,2020-08-06 23:16:27,2020-08-06,23,EEN,EUROPA LEAGUE MAGAZINE,1690,147325,21.9,22.8,0.0,0.0,0,14,6.3,0.0,0,3,0,zomer
29278,29278,2020-10-26 19:00:05,2020-10-26,19,EEN,HET 7 UUR-JOURNAAL,2714,1296824,10.6,6.8,0.0,0.0,3,81,20.1,0.0,0,0,0,herfst
7123,7123,2017-09-26 18:59:46,2017-09-26,18,VTM,NIEUWS 19U VTM,3059,586310,17.4,17.2,0.0,0.0,1,42,8.7,3600.0,0,1,0,herfst
43532,43533,2022-10-11 18:59:37,2022-10-11,18,VTM,NIEUWS 19U VTM,3115,617035,14.3,13.3,0.0,0.0,3,83,3.1,813.5,0,1,0,herfst


Feature engineering

In [7]:
def lagFeatures(df, n):
  for i in range(1,n+1):
    df[f'KijkersLag{i}'] = df.sort_values('FullDate').groupby('Programma')['Kijkers'].shift(i).fillna(df.groupby('Programma')['Kijkers'].transform('mean'))
  return df

In [8]:
cleanData = lagFeatures(cleanData, 3)
cleanData['Kanaal'] = cleanData['Kanaal'].str.replace(' ', '_')

print(cleanData['Kanaal'].unique())

cleanData[cleanData['Programma'] == 'THUIS'][['FullDate', 'Kijkers', 'KijkersLag1', 'KijkersLag2', 'KijkersLag3']]


['EEN' 'VTM' 'CANVAS' 'VIER' 'VIJF' 'Q2' 'VITAYA' 'OP_12' 'ZES' 'LA_UNE'
 'RTL-TVI' 'TF1' 'AB3' 'KETNET' 'Canvas' 'CAZ' 'VTM2'
 'ELEVEN_PRO_LEAGUE_1_NL' 'VTM3' 'VTM4' 'PLAY_SPORTS_OPEN' 'PLAY4' 'PLAY5'
 'EUROSPORT_1_(NL)' 'PLAY6' 'VRT_1' 'VRT_CANVAS' 'VTM_GOLD'
 'DAZN_PRO_LEAGUE_1_(NL)']


,FullDate,Kijkers,KijkersLag1,KijkersLag2,KijkersLag3
40,2016-10-03 20:14:23,1268561,1.059928e+06,1.059928e+06,1.059928e+06
60,2016-10-04 20:09:27,1169791,1.268561e+06,1.059928e+06,1.059928e+06
80,2016-10-05 20:13:52,1244502,1.169791e+06,1.268561e+06,1.059928e+06
121,2016-10-07 19:57:34,1156477,1.244502e+06,1.169791e+06,1.268561e+06
180,2016-10-10 19:59:00,1315826,1.156477e+06,1.244502e+06,1.169791e+06
...,...,...,...,...,...
61265,2025-04-04 20:24:13,779244,1.018053e+06,9.942260e+05,1.066599e+06
61325,2025-04-07 20:14:35,869193,7.792440e+05,1.018053e+06,9.942260e+05
61345,2025-04-08 20:13:09,880767,8.691930e+05,7.792440e+05,1.018053e+06
61366,2025-04-09 20:19:53,802132,8.807670e+05,8.691930e+05,7.792440e+05


Kardinaliteiten

In [9]:
#kardinaliteiten voor betere toepassingen
for kolom in cleanData.columns:
  print(f'{kolom}: {cleanData[kolom].nunique()}')

Unnamed: 0: 61405
FullDate: 61318
date: 3051
hour: 21
Kanaal: 29
Programma: 5963
Lengte_sec: 6774
Kijkers: 58860
Temperatuur: 414
Gevoelstemp: 478
Regen: 62
Sneeuw: 25
Weercode: 13
Bewolking: 101
Windsnelheid: 456
Zonnenschijn: 3555
isFeestdag: 2
Weekdag: 7
isWeekend: 2
Seizoen: 4
KijkersLag1: 58998
KijkersLag2: 56714
KijkersLag3: 55143


One Hot Encoding

In [10]:
#one hot encoding voor lage cardinaliteit
oneHotEnc = OneHotEncoder(handle_unknown="ignore")
lageKardinaliteit = cleanData[[ 'hour','Kanaal', 'isFeestdag', 'Weekdag', 'Seizoen']]
oneHot = oneHotEnc.fit_transform(lageKardinaliteit)
dfOneHot = pd.DataFrame(oneHot.toarray(), 
                            columns=oneHotEnc.get_feature_names_out(), 
                            index=lageKardinaliteit.index)

cleanData = cleanData.drop(columns=['hour', 'Kanaal', 'isFeestdag', 'Weekdag', 'Seizoen'])
cleanData = pd.concat([cleanData, dfOneHot], axis=1)

cleanData.sample(5)

,Unnamed: 0,FullDate,date,Programma,Lengte_sec,Kijkers,Temperatuur,Gevoelstemp,Regen,Sneeuw,...,Weekdag_1,Weekdag_2,Weekdag_3,Weekdag_4,Weekdag_5,Weekdag_6,Seizoen_herfst,Seizoen_lente,Seizoen_winter,Seizoen_zomer
57498,57501,2024-09-23 18:24:23,2024-09-23,MILO,1398,267503,17.5,16.5,0.1,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0
38177,38177,2022-01-14 16:32:05,2022-01-14,BRAIN BOXES,2663,145155,4.2,2.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
41229,41230,2022-06-11 12:35:56,2022-06-11,DE MARKT,1411,140187,19.5,19.5,0.0,0.0,...,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0
26996,26996,2020-06-18 21:33:29,2020-06-18,DE KUST IS VEILIG,1991,204114,18.2,19.0,0.0,0.0,...,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
41726,41727,2022-07-09 23:49:47,2022-07-09,SECTION DE RECHERCHES,2805,85308,14.6,13.3,0.0,0.0,...,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0


In [11]:
#opslaan bij models
modelsDir = os.getcwd()
modelFile = os.path.join(modelsDir, ".",
                          "models", "oneHotEncoder.pkl")
os.makedirs(os.path.dirname(modelFile), exist_ok=True)
with open(modelFile, 'wb') as m:
  pickle.dump(oneHotEnc, m)

Target Encoding

In [12]:
#target encoding voor medium kardinaliteiten
targetEncoding = TargetEncoder()
medKardinaliteit = cleanData[['date', 'Programma', 'Lengte_sec', 'Temperatuur', 'Gevoelstemp', 'Regen', 'Bewolking', 'Windsnelheid', 'Zonnenschijn']]
#verdere feature engineering op vorig model
target = targetEncoding.fit_transform(medKardinaliteit, cleanData['Programma'])
target.sample(10)

,date,Programma,Lengte_sec,Temperatuur,Gevoelstemp,Regen,Bewolking,Windsnelheid,Zonnenschijn
10084,2244.375912,2554.000000,3158,4.4,-1.5,0.0,50,21.8,3600.0
4869,2235.425912,1676.000000,2060,21.0,17.5,0.0,47,25.4,3600.0
30873,2133.225912,1401.494741,2908,0.9,-4.2,0.1,100,20.6,0.0
52387,2442.966512,5139.045129,3815,7.0,3.1,0.2,100,18.3,0.0
9128,2175.475912,1749.000000,957,5.8,1.4,0.0,16,20.7,0.0
43246,2024.875912,3419.000000,2066,10.9,8.0,0.5,98,17.1,0.0
20103,2332.800912,2236.999446,1780,17.2,13.8,0.0,46,17.5,3600.0
3533,2284.700912,1749.000000,923,12.3,9.6,0.0,92,12.8,0.0
41257,2404.619077,2549.000000,2201,19.5,18.2,0.0,100,16.0,3600.0
2943,2297.425912,589.000000,1574,6.0,1.7,0.0,11,16.5,3600.0


In [13]:
cleanData = cleanData.drop(columns=['date', 'Programma', 'Lengte_sec', 'Temperatuur', 'Gevoelstemp', 'Regen', 'Bewolking', 'Windsnelheid', 'Zonnenschijn'])
cleanData = pd.concat([cleanData, target], axis=1)
cleanData.sample(5)

,Unnamed: 0,FullDate,Kijkers,Sneeuw,Weercode,isWeekend,KijkersLag1,KijkersLag2,KijkersLag3,hour_0,...,Seizoen_zomer,date,Programma,Lengte_sec,Temperatuur,Gevoelstemp,Regen,Bewolking,Windsnelheid,Zonnenschijn
34199,34199,2021-06-30 19:51:39,681496,0.0,53,0,685445.0,549635.0,572349.0,0.0,...,1.0,2213.750912,1425.000000,2194,14.0,12.5,0.5,100,14.4,0.0
32735,32735,2021-04-17 20:27:12,154826,0.0,0,1,262110.0,208468.0,208468.0,0.0,...,0.0,2026.300912,2191.114246,6879,7.4,3.3,0.0,16,15.9,3600.0
28489,28489,2020-09-16 22:42:19,365621,0.0,1,0,263951.0,301686.0,357658.0,0.0,...,1.0,2186.975912,1749.000000,1023,18.6,17.8,0.0,20,19.2,0.0
30684,30684,2021-01-05 20:06:32,670990,0.0,3,0,720107.0,664295.0,577678.0,0.0,...,0.0,2155.075912,1403.000000,1543,2.5,-0.7,0.0,100,9.5,0.0
28863,28863,2020-10-05 20:07:17,634203,0.0,3,0,657403.0,621477.0,636337.0,0.0,...,0.0,2248.825912,1403.000000,1536,12.8,10.4,0.0,88,18.3,0.0


Verbanden

In [14]:
dataOneHotTarget = cleanData.select_dtypes(include=[np.number])

verband = dataOneHotTarget.corr()

verband = verband['Kijkers'].abs().sort_values(ascending=False)

print(f'{verband.head(10)}')
dataOneHotTarget.shape

Kijkers        1.000000
KijkersLag1    0.894582
KijkersLag2    0.893195
KijkersLag3    0.875979
hour_19        0.386184
Kanaal_EEN     0.333420
hour_22        0.184308
hour_17        0.176407
hour_20        0.176109
Gevoelstemp    0.172495
Name: Kijkers, dtype: float64


(61405, 80)

In [15]:
# Eerst omzetten naar strings, daarna komma's verwijderen, en uiteindelijk naar floats
for col in ['KijkersLag1', 'KijkersLag2', 'KijkersLag3', 'Programma']:
    cleanData[col] = cleanData[col].astype(str).str.replace(',', '').astype(float)

# Print het resultaat
print(cleanData[['Kijkers', 'KijkersLag1', 'KijkersLag2', 'KijkersLag3', 'Programma']].head(10))


   Kijkers    KijkersLag1    KijkersLag2    KijkersLag3    Programma
0   721850  892001.072096  892001.072096  892001.072096  1677.000000
1   709606  628487.089971  628487.089971  628487.089971  1425.000000
2   548239  478421.179487  478421.179487  478421.179487  4807.106007
3   523610  791862.682708  791862.682708  791862.682708  1878.000000
4   496216  491224.740741  491224.740741  491224.740741   779.215215
5   447427  351454.000000  351454.000000  351454.000000  2411.204623
6   424041  585853.225914  585853.225914  585853.225914  2554.000000
7   369066  408691.165611  408691.165611  408691.165611  1676.000000
8   368549  336230.500000  336230.500000  336230.500000  2885.510725
9   360544  332870.588235  332870.588235  332870.588235  3291.702083


In [16]:
#opslaan bij models
targetModelDir = os.getcwd()
targetModelPath = os.path.join(targetModelDir, ".",
                          "models", "oneHotTarget.pkl")
os.makedirs(os.path.dirname(targetModelPath), exist_ok=True)
with open(targetModelPath, 'wb') as m:
  pickle.dump(targetEncoding, m)

Model tests

In [13]:
def test_models(models, X, y, cv=10, scoring='neg_mean_absolute_error'):
    res = {}
    for m in models:
        model = type(m).__name__
        scores = cross_val_score(m, X, y, cv=cv, scoring=scoring)
        res[model] = {
            'scores': scores,
            'mean_score': scores.mean()
        }
        print(f"Scores {model}\n{scores}\nMAE = {scores.mean()}")
    return res

Lasso Regression

In [17]:
models = [
    RandomForestRegressor(verbose=True), Lasso()
]

# Zorg ervoor dat X en y goed geschaald zijn
scaler = StandardScaler()
X = dataOneHotTarget.drop(columns=['Kijkers', 'Unnamed: 0'])
X_scaled = scaler.fit_transform(X)
y = dataOneHotTarget['Kijkers']

In [ ]:
# Run de test
results = test_models(models, X_scaled, y)

RandomForest heeft betere scores:
  
  RandomForest
  MAE = -52082.47210795064

  Lasso
  MAE = -64622.292775108515

Naar csv schrijven

In [19]:
dir = os.getcwd()
pathFile = os.path.join(dir, "./data csv/oneHotTarget.csv")

os.makedirs(os.path.dirname(pathFile), exist_ok=True)

dataOneHotTarget.to_csv(pathFile)


keuze maken voor model

In [16]:
#trainset en testset maken
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=True)

def scoreModel(model,X,y, cv=5):
    print(f"Model: {model.__class__.__name__}")
    start = time.time()
        
    # MAE
    mae_scores = cross_val_score(model, X, y, cv=cv, scoring='neg_mean_absolute_error')
    mae = -np.mean(mae_scores)

    # MAPE
    mape_scores = cross_val_score(model, X, y, cv=cv, scoring='neg_mean_absolute_percentage_error')
    mape = -np.mean(mape_scores) * 100

    duration = time.time() - start
    min, sec = divmod(duration, 60)

    print(f"MAE:  {mae:.4f}")
    print(f"MAPE: {mape:.2f}%")
    print(f"Tijd: {int(min)}m {sec:.1f}s")

    return {'Model': model.__class__.__name__, 'MAE': mae, 'MAPE': mape, 'Tijd': duration}
def startModel(models, X, y, cv=5):
    results = []
    for model in models:
        result = scoreModel(model, X, y, cv=cv)
        results.append(result)
    return results

In [17]:
linear = LinearRegression()
decision = DecisionTreeRegressor()
#randomForest = RandomForestRegressor()
gradientBoosting = GradientBoostingRegressor()
xgb = XGBRegressor()
extraTrees = ExtraTreesRegressor()
histGB = HistGradientBoostingRegressor()
lgbm = LGBMRegressor()

startModel([linear, decision, gradientBoosting, xgb, extraTrees, histGB, lgbm], X, y)


Model: LinearRegression
MAE:  66486.7244
MAPE: 20.31%
Tijd: 0m 1.6s
Model: DecisionTreeRegressor


KeyboardInterrupt: 

# Best models:

## Model: ExtraTreesRegressor

MAE: 54860.2905

MAPE: 16.27%

Tijd: 11m 50.4s

## Model: HistGradientBoostingRegressor

MAE:  55206.2561

MAPE: 16.29%

Tijd: 0m 11.2s

## Model: LGBMRegressor

MAE:  55006.0849

MAPE: 16.23%

Tijd: 0m 3.4s

## RandomForest er niet bij want 18m+ traintijd voor slechts 200 MAE verschil met XGB


ExtraTrees: Beste nauwkeurigheid maar lange training tijd

HistGradientBoosting en LGBM: Veel sneller met minimale performance drop

Verbeteren van de top 2 modellen

In [24]:
warnings.filterwarnings("ignore")

#parameters
model_param = {
    # 'LGBMRegressor': {
    #     'model': LGBMRegressor(),
    #     'params': {
    #         'n_estimators': randint(100, 1000),
    #         'learning_rate': uniform(0.01, 0.3),
    #         'max_depth': randint(3, 15),
    #         'min_child_weight': randint(1, 10),
    #         'subsample': uniform(0.6, 0.4),
    #         'colsample_bytree': uniform(0.6, 0.4),
    #         'gamma': uniform(0, 0.5),
    #         'reg_lambda': uniform(0.1, 10),
    #         'reg_alpha': uniform(0, 10)
    #     }
    # },
    # 'HistGradient': {
    #     'model': HistGradientBoostingRegressor(),
    #     'params': {
    #         'learning_rate': uniform(0.01, 0.3),
    #         'max_iter': randint(100, 1000),
    #         'max_depth': randint(3, 15),
    #         'l2_regularization': uniform(0.1, 10),
    #         'min_samples_leaf': randint(1, 10),
    #         'max_leaf_nodes': randint(20, 100)
    #     }
    # },
    'LightGBM': {
        'model': LGBMRegressor(),
        'params': {
            'n_estimators': [850],
            'num_leaves': [31, 50],
            'learning_rate': [0.01, 0.1],
            'n_estimators': [100, 200],
            'max_depth': [-1, 5],
            'min_child_samples': [10, 20],
            'subsample': [0.6, 0.8],
            'colsample_bytree': [0.6, 0.8],
            'reg_alpha': [0, 0.1],
            'reg_lambda': [0, 0.1]
        }
    }
}

extra = ExtraTreesRegressor()
hist = HistGradientBoostingRegressor()
xgb = XGBRegressor()
lgbm = LGBMRegressor()


Random Search

In [20]:
for name, config in model_param.items():
    print(f"Model: {name}")
    model = config['model']
    param = config['params']
    
    rsLoop = RandomizedSearchCV(
        estimator=model,
        param_distributions=param,
        n_iter=250,
        cv=5,
        scoring='neg_mean_absolute_error',
        n_jobs=-1,
        verbose=2
    )
    
    start = time.time()
    rsLoop.fit(X_train, y_train)
    duration = time.time() - start

    print(f"Tijd {name}: {duration:.2f}s")
    print(f"Beste param {name}: {rsLoop.best_params_}")
    print(f"Beste MAE: {rsLoop.best_score_:.4f}")

Model: LGBMRegressor
Fitting 5 folds for each of 250 candidates, totalling 1250 fits
[LightGBM] [Warning] Unknown parameter: gamma
[LightGBM] [Warning] Unknown parameter: gamma
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.003941 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 2817
[LightGBM] [Info] Number of data points in the train set: 49124, number of used features: 59
[LightGBM] [Info] Start training from score 446116.245705
Tijd LGBMRegressor: 914.07s
Beste param LGBMRegressor: {'colsample_bytree': np.float64(0.6159927582454119), 'gamma': np.float64(0.4252234490130524), 'learning_rate': np.float64(0.0797065159450008), 'max_depth': 11, 'min_child_weight': 6, 'n_estimators': 892, 'reg_alpha': np.float64(9.926651888227628), 'reg_lambda': np.float64(2.3831325944838744), 'subsample': np.float64(0.6189602366096897)}
Beste MAE: -50255.

In [21]:
best_params = rsLoop.best_params_
tuned_model = rsLoop.best_estimator_

y_pred = tuned_model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mape = mean_absolute_percentage_error(y_test, y_pred)
accuracy = 1 - mape

print(f'Mean absolute error: {mae}')
print(f'Mean absolute percentage error: {mape}')
print(f'Accuracy: {accuracy}')

Mean absolute error: 50662.802550964676
Mean absolute percentage error: 0.14810066764575672
Accuracy: 0.8518993323542433


In [ ]:

rsExtra = RandomizedSearchCV(
  estimator=extra,
  param_distributions=model_param['ExtraTrees']['params'],
  n_iter=150,
  cv=5,
  scoring='neg_mean_absolute_error',
  n_jobs=-1,
  verbose=2
)
    
start = time.time()
rsExtra.fit(X_train, y_train)
duration = time.time() - start

print(f"Tijd: {duration:.2f}s")
print(f"Beste param: {rsExtra.best_params_}")
print(f"Beste MAE: {rsExtra.best_score_:.4f}")

Fitting 5 folds for each of 100 candidates, totalling 500 fits
Tijd: 2857.67s
Beste param: {'max_depth': 9, 'max_features': np.float64(0.9124558702471834), 'min_samples_leaf': 2, 'min_samples_split': 4, 'n_estimators': 742}
Beste MAE: -54224.2000


### XGBoost
Tijd XGBoost: 1334.04s
Beste param XGBoost: {'colsample_bytree': np.float64(0.6550054872406584), 'gamma': np.float64(0.08367508771505777), 'learning_rate': np.float64(0.04567558428183022), 'max_depth': 8, 'min_child_weight': 2, 'n_estimators': 609, 'reg_alpha': np.float64(6.924415940372761), 'reg_lambda': np.float64(0.2620475254770348), 'subsample': np.float64(0.970829180420445)}
Beste MAE: -46309.9074
### HistGradient
Tijd HistGradient: 1145.61s
Beste param HistGradient: {'l2_regularization': np.float64(6.796320445820958), 'learning_rate': np.float64(0.02650163315092701), 'max_depth': 9, 'max_iter': 770, 'max_leaf_nodes': 96, 'min_samples_leaf': 4}
Beste MAE: -47550.1241
### ExtraTree
Tijd: 2857.67s
Beste param: {'max_depth': 9, 'max_features': np.float64(0.9124558702471834), 'min_samples_leaf': 2, 'min_samples_split': 4, 'n_estimators': 742}
Beste MAE: -54224.2000
### LightGBM
Tijd LightGBM: 448.79s
Beste param LightGBM: {'subsample': 1.0, 'reg_lambda': 0, 'reg_alpha': 0.5, 'num_leaves': 50, 'n_estimators': 500, 'min_child_samples': 10, 'max_depth': 10, 'learning_rate': 0.1, 'colsample_bytree': 0.8}
Beste MAE: -47344.7920

In [27]:
def run_optuna(model_name):
    def objective(trial):
        # if model_name == 'ExtraTrees':
        #     model = ExtraTreesRegressor(
        #         n_estimators=trial.suggest_int('n_estimators', 100, 500),
        #         max_depth=trial.suggest_int('max_depth', 5, 30),
        #         min_samples_split=trial.suggest_int('min_samples_split', 2, 10),
        #         min_samples_leaf=trial.suggest_int('min_samples_leaf', 1, 10),
        #         random_state=42
        #     )
        # if model_name == 'HistGradient':
        #     model = HistGradientBoostingRegressor(
        #         learning_rate=trial.suggest_float('learning_rate', 0.01, 0.3),
        #         max_iter=trial.suggest_int('max_iter', 200, 500),
        #         max_depth=trial.suggest_int('max_depth', 3, 15),
        #         max_leaf_nodes=trial.suggest_int('max_leaf_nodes', 20, 100),
        #         random_state=42
        #     )
        # elif model_name == 'XGBoost':
        #     model = XGBRegressor(
        #         n_estimators=trial.suggest_int('n_estimators', 200, 800),
        #         max_depth=trial.suggest_int('max_depth', 10, 15),
        #         learning_rate=trial.suggest_float('learning_rate', 0.01, 0.2, log=True),
        #         subsample=trial.suggest_float('subsample', 0.5, 1.0),
        #         colsample_bytree=trial.suggest_float('colsample_bytree', 0.5, 1.0),
        #         reg_alpha=trial.suggest_float('reg_alpha', 0.0, 1.0),
        #         reg_lambda=trial.suggest_float('reg_lambda', 0.0, 1.0),
        #         min_child_weight=trial.suggest_int('min_child_weight', 1, 10),
        #         gamma=trial.suggest_float('gamma', 0, 5),
        #         random_state=42,
        #         n_jobs=-1
        # )
        if model_name == 'LightGBM':
            model = LGBMRegressor(
                n_estimators=trial.suggest_int('n_estimators', 650, 800),
                max_depth=trial.suggest_int('max_depth', 10, 14),
                num_leaves=trial.suggest_int('num_leaves', 180, 230),
                learning_rate=trial.suggest_float('learning_rate', 0.025, 0.045, log=True),
                subsample=trial.suggest_float('subsample', 0.92, 0.98),
                colsample_bytree=trial.suggest_float('colsample_bytree', 0.58, 0.65),
                reg_alpha=trial.suggest_float('reg_alpha', 0.02, 0.07),
                reg_lambda=trial.suggest_float('reg_lambda', 0.02, 0.07),
                min_child_samples=trial.suggest_int('min_child_samples', 17, 23),
                random_state=42,
                n_jobs=-1,
                verbosity=-1
            )
        model.fit(X_train, y_train)
        trial.set_user_attr("model", model)
        pred = model.predict(X_test)
        return mean_absolute_error(y_test, pred)
    return objective

In [28]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=True)

for name, config in model_param.items():
  print(f"Optuna {name}")
  optres = optuna.create_study(direction="minimize")
  optres.optimize(run_optuna(name), n_trials=500)

  start = time.time()
  duration = time.time() - start

  print(f"Tijd: {duration:.2f}s")
  print(f"Beste params {name}: ", optres.best_params)
  print(f"Beste MAE {name}: ", optres.best_value)

[I 2025-05-15 16:59:48,747] A new study created in memory with name: no-name-3016358b-466d-414f-8296-b55a7f181270


Optuna LightGBM


[I 2025-05-15 16:59:55,428] Trial 0 finished with value: 49120.55153279656 and parameters: {'n_estimators': 768, 'max_depth': 14, 'num_leaves': 180, 'learning_rate': 0.033110721897355345, 'subsample': 0.9390092078048248, 'colsample_bytree': 0.6456533111674649, 'reg_alpha': 0.046155408040214244, 'reg_lambda': 0.02166863931766721, 'min_child_samples': 18}. Best is trial 0 with value: 49120.55153279656.
[I 2025-05-15 17:00:01,511] Trial 1 finished with value: 49352.5753245293 and parameters: {'n_estimators': 685, 'max_depth': 11, 'num_leaves': 217, 'learning_rate': 0.042275968234874034, 'subsample': 0.9593833743701587, 'colsample_bytree': 0.6229402137867456, 'reg_alpha': 0.039799232275720894, 'reg_lambda': 0.0662249374941952, 'min_child_samples': 17}. Best is trial 0 with value: 49120.55153279656.
[I 2025-05-15 17:00:06,572] Trial 2 finished with value: 49427.37268126975 and parameters: {'n_estimators': 757, 'max_depth': 10, 'num_leaves': 193, 'learning_rate': 0.0426558552596221, 'subsamp

Tijd: 0.00s
Beste params LightGBM:  {'n_estimators': 766, 'max_depth': 12, 'num_leaves': 194, 'learning_rate': 0.0331229115457035, 'subsample': 0.9711304779976127, 'colsample_bytree': 0.6407629168857206, 'reg_alpha': 0.06966464205277899, 'reg_lambda': 0.043201120991116455, 'min_child_samples': 17}
Beste MAE LightGBM:  48952.449281085166


extratrees word verder genegeerd door de zeer lange trainingtijden en slechte finetuning resultaten, hierdoor heb ik dit vervamgen met LGBM doordat dit gelijkaardig zou moetten testen als XGB op mijn model

Optuna XGBoost

Tijd: 0.00s
Beste params XGBoost:  {'n_estimators': 702, 'max_depth': 11, 'learning_rate': 0.03341587956729562, 'subsample': 0.9067542838662811, 'colsample_bytree': 0.5427907826581247, 'reg_alpha': 0.059647169243809806, 'reg_lambda': 0.6797096185328306, 'min_child_weight': 8, 'gamma': 2.218523626972447}
[I 2025-05-15 13:41:09,427] A new study created in memory with name: no-name-c866fb08-43b3-4960-8232-e77463cea1d3
Beste MAE XGBoost:  49130.40625

Optuna HistGradient

Tijd: 0.00s
Beste params HistGradient:  {'learning_rate': 0.07132678642273478, 'max_iter': 450, 'max_depth': 11, 'max_leaf_nodes': 73}
Beste MAE HistGradient:  50422.51862966884

Optuna LightGBM

Beste params LightGBM:  {'n_estimators': 725, 'max_depth': 13, 'num_leaves': 203, 'learning_rate': 0.03441489414795745, 'subsample': 0.9607189226837076, 'colsample_bytree': 0.600584267464685, 'reg_alpha': 0.0367710312215926, 'reg_lambda': 0.05482079934133707, 'min_child_samples': 19}
Beste MAE LightGBM:  48961.120502551305

In [26]:
print(optres.best_trial.user_attrs)


{'model': LGBMRegressor(colsample_bytree=0.600584267464685,
              learning_rate=0.03441489414795745, max_depth=13,
              min_child_samples=19, n_estimators=725, n_jobs=-1, num_leaves=203,
              random_state=42, reg_alpha=0.0367710312215926,
              reg_lambda=0.05482079934133707, subsample=0.9607189226837076,
              verbosity=-1)}


In [27]:
best_params = optres.best_params
tuned_model = optres.best_trial.user_attrs['model']

y_pred = tuned_model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mape = mean_absolute_percentage_error(y_test, y_pred)
accuracy = 1 - mape

print(f'Mean absolute error: {mae}')
print(f'Mean absolute percentage error: {mape}')
print(f'Accuracy: {accuracy}')

Mean absolute error: 48961.120502551305
Mean absolute percentage error: 0.14284014800967923
Accuracy: 0.8571598519903207


In [29]:
besteModel = optres.best_trial
#opslaan bij models
lgbmModelDir = os.getcwd()
lgbmModelPath = os.path.join(lgbmModelDir, ".",
                          "models", "lightGBM.pkl")
os.makedirs(os.path.dirname(lgbmModelPath), exist_ok=True)
with open(lgbmModelPath, 'wb') as m:
    pickle.dump(besteModel.user_attrs['model'], m)

## Best so far
### LightGBM
Mean absolute error: 45337.28982557999
Mean absolute percentage error: 0.1334235812403486
Accuracy: 0.8665764187596514

GridsearchCV

In [32]:

#parameters
model_param = {
    'XGBoost': {
        'model': XGBRegressor(),
        'params': {
            'n_estimators': [850],
            'learning_rate': [0.025, 0.03],
            'max_depth': [8, 9],
            'min_child_weight': [1,2],
            'subsample': [1.0],
            'colsample_bytree': [0.8],
            'gamma': [0.035, 0.04],
            'reg_lambda': [3],
            'reg_alpha': range(0, 5, 1)
        }
    },
    'HistGradient': {
        'model': HistGradientBoostingRegressor(),
        'params': {
            'n_estimators': [850],
            'learning_rate': [0.01, 0.1],
            'max_iter': [100, 200],
            'max_depth': [3, 5],
            'l2_regularization': [0.1, 1],
            'min_samples_leaf': [1, 2],
            'max_leaf_nodes': [20, 40]
        }
    },
    'LightGBM': {
        'model': LGBMRegressor(),
        'params': {
            'n_estimators': [850],
            'num_leaves': [31, 50],
            'learning_rate': [0.01, 0.1],
            'n_estimators': [100, 200],
            'max_depth': [-1, 5],
            'min_child_samples': [10, 20],
            'subsample': [0.6, 0.8],
            'colsample_bytree': [0.6, 0.8],
            'reg_alpha': [0, 0.1],
            'reg_lambda': [0, 0.1]
        }
    }
}

hist = HistGradientBoostingRegressor()
xgb = XGBRegressor()
lgbm = LGBMRegressor()


In [42]:
for name, config in model_param.items():
    print(f"Model: {name}")
    model = config['model']
    param = config['params']
    
    gs = GridSearchCV(
        estimator=model,
        param_grid=param,
        cv=5,
        scoring='neg_mean_absolute_error',
        n_jobs=-1,
        verbose=2
    )
    
    start = time.time()
    gs.fit(X_train, y_train)
    duration = time.time() - start

    print(f"Tijd {name}: {duration:.2f}s")
    print(f"Beste param {name}: {gs.best_params_}")
    print(f"Beste MAE: {gs.best_score_:.4f}")

Model: XGBoost
Fitting 5 folds for each of 80 candidates, totalling 400 fits


KeyboardInterrupt: 

Resultaten tuned model

In [75]:
best_params = gs.best_params_
tuned_model = gs.best_estimator_

y_pred = tuned_model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mape = mean_absolute_percentage_error(y_test, y_pred)
accuracy = 1 - mape

print(f'Mean absolute error: {mae}')
print(f'Mean absolute percentage error: {mape}')
print(f'Accuracy: {accuracy}')

NameError: name 'gs' is not defined

### model opslaan met beste parameters

In [27]:
besteModel = LGBMRegressor(
    n_estimators=788,
    max_depth=0,
    num_leaves=213,
    learning_rate=0.027352468420388434,
    subsample=0.8257111258276627,
    colsample_bytree=0.7709277343834496,
    reg_alpha=0.7512362258288539,
    reg_lambda=0.23466677100487698,
    min_child_samples=8,
    random_state=42,
    n_jobs=-1,
    verbosity=-1
)

besteModel.fit(X_train, y_train)
pred = besteModel.predict(X_test)
mae = mean_absolute_error(y_test, pred)
print(f"MAE: {mae:.4f}")

MAE: 48910.7069


In [78]:
#opslaan bij models
lgbmModelDir = os.getcwd()
lgbmModelPath = os.path.join(lgbmModelDir, ".",
                          "models", "lightGBM.pkl")
os.makedirs(os.path.dirname(lgbmModelPath), exist_ok=True)
with open(lgbmModelPath, 'wb') as m:
    pickle.dump(besteModel, m)